### PR TITLE
Enable workload identity for event stream

### PR DESCRIPTION
**Pull request by mismithhisler**
### Description

Enables Workload Identities to subscribe to event streams.  Previously, the event broker used custom authentication logic that could only accept the secretID from an ACL Token.  This change refactors the event broker to use an authentication callback for handling authentication while subscribing and after various ACL events.  This allows us to inject the standardized `Authenticator` logic into the event broker.

### Testing & Reproduction steps

Enable ACL's, create the appropriate policy, associate it with a workload, and use the WI as the Bearer token when subscribing to the events stream.

### Links

Fixes [GH #18186](https://github.com/hashicorp/nomad/issues/18186)

### Contributor Checklist
- [x] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [x] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [x] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
- [ ] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [ ] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository.

### CODE REVIEW

File: nomad/event_endpoint.go [hunk 1]
```diff
@@ -213,3 +224,47 @@ func handleJsonResultError(err error, code *int64, encoder *codec.Encoder) {
...
+		switch topic {
+		case structs.TopicDeployment,
+			structs.TopicEvaluation,
+			structs.TopicAllocation,
+			structs.TopicJob,
+			structs.TopicService:
+			if ok := aclObj.AllowNsOp(namespace, acl.NamespaceCapabilityReadJob); !ok {
+				return structs.ErrPermissionDenied
+			}
+		case structs.TopicHostVolume:
```
Comment (mismithhisler) [inline nomad/event_endpoint.go:239]: I don't see subscribing to `HostVolume` or `CSI Read Volume` in the [events api docs](https://developer.hashicorp.com/nomad/api-docs/events).  Should we remove this, or update the docs to include these options?

File: nomad/stream/event_broker_test.go
Comment (mismithhisler) [inline nomad/stream/event_broker_test.go:1]: A lot of tests that covered the custom authentication logic have been removed from here, as we are now using the `Authenticator` logic via the authentication callback.

File: nomad/event_endpoint.go (hunk 1 above)
Comment (tgross) [reply nomad/event_endpoint.go:239]: These are added on main but not a release build.

File: nomad/event_endpoint.go [hunk 2]
```diff
@@ -61,6 +62,16 @@ func (e *Event) stream(conn io.ReadWriteCloser) {
 		Topics:    args.Topics,
 		Index:     uint64(args.Index),
 		Namespace: args.Namespace,
+		Authenticate: func() error {
+			if err := e.srv.Authenticate(nil, &args); err != nil {
+				return err
+			}
+			acl, err := e.srv.ResolveACL(&args)
```
Comment (jrasell) [inline nomad/event_endpoint.go:69]: `acl` shadows an import name, could we change this variable name?

File: nomad/event_endpoint.go [hunk 3]
```diff
@@ -74,21 +85,21 @@ func (e *Event) stream(conn io.ReadWriteCloser) {
...
 	if subErr != nil {
 		handleJsonResultError(subErr, pointer.Of(int64(500)), encoder)
 		return
 	}
 	defer subscription.Unsubscribe()
 
+	// because we have authenticated, the identity will be set, so extract expiration time
+	var exp time.Time
+	if c := args.GetIdentity().GetClaims(); c != nil {
+		exp = c.Expiry.Time()
```
Comment (jrasell) [inline nomad/event_endpoint.go:98]: Is there any possibility `Expiry` can be nil?

File: nomad/event_endpoint.go [hunk 4]
```diff
@@ -213,3 +224,47 @@ func handleJsonResultError(err error, code *int64, encoder *codec.Encoder) {
 		Error: structs.NewRpcError(err, code),
 	})
 }
+
+func validateACL(namespace string, topics map[structs.Topic][]string, aclObj *acl.ACL) error {
```
Comment (jrasell) [inline nomad/event_endpoint.go:228]: Is it worth adding a table-driven test to cover the possibilities of this function? If think our current tests only cover a portion of the function.

**Review by jrasell — COMMENTED** [review]
LGTM, thanks @mismithhisler!

I think the PR still needs a changelog entry and I added a couple of inline queries, so I'll comment as my approval will likely get removed with at least the changelog commit.

File: nomad/event_endpoint.go (hunk 3 above)
Comment (mismithhisler) [reply nomad/event_endpoint.go:98]: It's probably possible, but the `Time()` method does a nil check on Expiry.  If you want, I can add a nil check here just to be defensive?

File: nomad/event_endpoint.go (hunk 4 above)
Comment (mismithhisler) [reply nomad/event_endpoint.go:228]: Yes I think it is.  I originally was confused about why these untested portions were even present until @tgross they are new code added for 1.10.

I'll refactor the existing test to cover this better but just need to be careful to do it in a way that doesn't make backporting too painful.

Comment (tgross) [reply nomad/event_endpoint.go:228]: Except for topic-to-policy mapping, mostly this function is covered by `acl` package tests. So if we're going to add tests here, the primary value would be in making sure we're testing that mapping against corner cases like wildcard namespaces.

Comment (mismithhisler) [reply nomad/event_endpoint.go:228]: My thought here was not to test the internals of `AllowNsOp` (although I see the overlap) but make sure that this validation function correctly matches what's listed in the required ACL section of the [event stream](https://developer.hashicorp.com/nomad/api-docs/events#event-stream) docs.

Let me know if you think that's too much overlap with the `acl` package tests

Comment (tgross) [reply nomad/event_endpoint.go:228]: Makes sense to me!

**Review by jrasell — APPROVED** [review]
